### PR TITLE
[unified build] build changes to make cpu unified build working.

### PR DESCRIPTION
**Pull request by Yangqing**
Properly annotated all apis for cpu front. Checked with cmake using

cmake -DUSE_ATEN=ON -DUSE_CUDA=OFF -DBUILD_ATEN=ON

and resulting libcaffe2.so has about 11k symbols.

### CODE REVIEW

File: caffe2/CMakeLists.txt
Comment (Yangqing) [inline caffe2/CMakeLists.txt]: Note: this line is here just for testing. While we should eventually turn it on, this diff is not ready to turn it fully on yet.

File: aten/src/ATen/core/Macros.h
Comment (orionr) [inline aten/src/ATen/core/Macros.h]: Do we need to do the GNU version checks and all that business like

https://github.com/pytorch/pytorch/blob/master/caffe2/core/common.h#L117-L121

? If not, we should likely simplify both.

File: caffe2/core/registry.h
Comment (orionr) [inline caffe2/core/registry.h]: So with the `CAFFE2_EXPORT` I'm assuming this only gets called in `.cc` or `.cpp` files? Just trying to understand the choice.

**Review by orionr — APPROVED** [review]
Awesome! A few questions and comments. Also a real build failure at https://ci.pytorch.org/jenkins/job/caffe2-builds/job/py2-cuda9.0-cudnn7-windows-build/11013/console

File: aten/src/ATen/core/Macros.h
Comment (Yangqing) [reply aten/src/ATen/core/Macros.h]: I found that it seems to be working across all our builds, so I am basically only testing gnuc but not version.

File: caffe2/core/registry.h
Comment (Yangqing) [reply caffe2/core/registry.h]: Ah yeah, so basically, the choice of API (which at build time adapts to IMPORT or EXPORT), EXPORT and IMPORT are as follows:

(1) for function and class definitions in header files: always use API. This will then be routed at compile time: when we are building the main library, because we set CAFFE2_BUILD_MAIN_LIB, it routes us to EXPORT, allowing the definition in the .cc files to be exported. When we are using the headers in a dependent library (someone that uses Caffe2/pytorch), then because CAFFE2_BUILD_MAIN_LIB is not set, it routes to IMPORT, which tells the compiler the function definitions are going to be elsewhere.

(2) in .cc files where we actually define the variables (global variables, static variables and functions), explicitly using EXPORT is safer because it avoid programming error.

(3) in header files when defining macros, and if the macro is going to be used in cc files for global variable definition (such as this line) - explicitly use EXPORT. The reason is that, such lines may be used in dependent libraries too (imagine a dependent library defining a new registry for a new device) - and when they do so, they want to explicitly EXPORT the registry.

Comment (ezyang) [reply caffe2/core/registry.h]: Put it in a comment?

**Comment by Yangqing** [comment]
Windows build failure identified as a different cause - landing to unblock other PRs since this is a bit big.

File: caffe2/core/registry.h
Comment (Yangqing) [reply caffe2/core/registry.h]: @ezyang Thanks, will do - maybe after we merge all the macros... Right now it is kinda everywhere, and AT_API doesn't have export/import equivalent yet.

**Comment by orionr** [comment]
Oops - backing out my change. One sec.

**Pull request closed without merging**
